### ISSUE TITLE
Tab completion incorrectly completes the wrong file when starting with the end of the filename

### ISSUE THREAD

**Issue by protist** [zimfw/zimfw#212]
Please check the existing issues to make sure you're not duplicating a report.

For bug reports, please provide the following information:

In a terminal, run `zmanage info` and paste the output below:

```
$ zmanage info
zsh: command not found: zmanage
```
But I'm running the zim master via Arch Linux AUR's zsh-zim-git: 9eaba9a

- gist of .zimrc: `zmodules=(completion)`
- gist of .zshrc: `source /usr/lib/zim/init.zsh`

Description
-----------
Tab completion incorrectly fully completes the wrong file when starting with the end of the filename.

Steps to reproduce
------------------
Tab completion works fine if starting from the start of the filename.
```
$ touch foo-bar foo-baz-bar
$ touch f<tab>
$ touch foo-ba<tab>
 -- file --
foo-bar      foo-baz-bar
```

However, if I start typing from the back (or middle) of the filename, it fails.
```
$ touch bar<tab>
$ touch foo-bar<tab>
           ^ cursor here
$ touch foo-bar
                ^ cursor here
```
I would expect the second tab to bring up the completion-selection menu. However, it immediately completes to the shorter version instead.

Another slightly different case is if I type the *entirety* of the shared string.
```
$ touch -bar<tab>
$ touch foo-bar
                ^ cursor here
```
In this case, I'm not even given the cursor in the middle, it jumps straight to the end.

**Comment by Eriner** [comment]
I think the current behavior makes sense. It won't bring up the tab complete menu in line two of your second code block because the test of "is valid file under cursor, or can we complete to one" passes on the first condition. This applies to the third code block as well. For sake of completion, you're expecting it to behave as below:
```command
$ touch foo-chem-bar && rm foo-bar
$ touch -bar<tab>
$ touch foo--bar
            ^ cursor here
 -- file --
foo-baz-bar   foo-chem-bar
```

Notice the removal of `foo-bar` and the addition of `foo-chem-bar`

Your example doesn't work because it completes that to the prefix of the possible options ("foo"), and then performs the "is this a valid entry" test when you hit tab the second time, which passes.

**Comment by protist** [comment]
Yes, your code block is what I'm expecting.

IMO merely passing "is this a valid entry" test shouldn't cause tab to move on, to the end of the line. Instead, I think it should be "is this the *only* valid entry". Otherwise, how would you ever get to select/complete `foo-baz-bar` in my second block?

In addition, the current behaviour still differs to starting at the beginning of the string. In the following example, the two files have the same beginning. After the second `<tab>`, it should pass "is this a valid entry" test, but *won't* auto-select it here (presumably because it doesn't pass "is this the *only* valid entry").
```
$ touch foo-bar foo-bar-baz
$ touch f<tab>
$ touch foo-bar<tab>
               ^ cursor here
$ touch foo-bar
 -- file --
foo-bar      foo-bar-baz
```

**Comment by Eriner** [comment]
Hm. I'm not sure what would need to be changed to meet your expectations of behavior of the shell. This is one of those quirks where I'm not sure where the "fix" needs to be.

**Comment by protist** [comment]
I'm not sure if you mean you are not sure what part of the code to change, or what end behaviour to change? If the latter, then my primary expectation would be as follows.

1) My main confusion is that tab should not auto-complete if there are other alternatives available. I feel like this behaviour is implied especially if the first tab places the cursor in the first ambiguous location, as per the second code block in the original post.

2) I guess this should only apply to middle-of-the-string completion. If the user starts typing at the beginning of the string, I suppose the default option should be to auto-complete the full string, even if there is another file that contains the visible "typed-out" string.

Thinking about it a bit more, that's probably the problem with the current implementation. As per the second code block above:

```
$ touch bar<tab>      # this tab auto-completes from the back/middle, as per (1)
$ touch foo-bar<tab>  # this tab has no knowledge of the previous completion, and completes as (2)
           ^ cursor here
$ touch foo-bar
                ^ cursor here
```

I'm not sure what the resolution is programmatically. Is there no way to "record" the type of auto-completion in the first line? Or could auto-completion behave differently if the cursor is located in the middle of the string, at an ambiguous location?

**Comment by Eriner** [comment]
I meant what zstyle or other options need to be changed to provide the behavior you expect.

**Comment by protist** [comment]
Oh. Right. Apologies for the misinterpretation.

**Comment by ericbn** [comment]
@protist, the [Completion and expansion](http://zsh.sourceforge.net/Guide/zshguide06.html#l145) title in Chapter 6 of the Z-Shell Guide has:

>     % ls
>     filename1
>     % ls filex
> Move the cursor to the `x` and hit tab. With `expand-or-complete` nothing happens; it's trying to complete a file called 'filex' --- or, with the option `COMPLETE_IN_WORD` set, it's trying to find a file whose name starts with 'file' and ends with 'x'. If you do
>     bindkey '^i' expand-or-complete-prefix
> and try the same experiment, you will find the whole thing is completed to 'filename1x', so that the 'x' was ignored, but not removed.

**TL;DR**: Typing `ls filex` and moving the cursor over the `x` yields different results depending on the zsh configuration used:

Configuration | Result
---|---
`bindkey '^i' expand-or-complete` | Tries to complete a file called 'filex'. **This is currently how zim is configured to work**
`bindkey '^i' expand-or-complete && setopt COMPLETE_IN_WORD` | Tries to find a file whose name starts with 'file' and ends with 'x'
`bindkey '^i' expand-or-complete-prefix` | Ignores the 'x' when completing, but doesn't remove it

@protist, it the behavior you're looking for between these 3 options here?

**Comment by ericbn** [comment]
@protist, using `setopt COMPLETE_IN_WORD` will give you the behavior you expect:

```
$ setopt COMPLETE_IN_WORD
$ touch foo-bar foo-baz-bar
$ touch bar<tab>
$ touch foo-bar<tab>
           ^ cursor here
$ touch foo-bar
              ^ cursor here
-- file --
foo-bar      foo-baz-bar
```

I think this is something good to have in zim as a sane default. @Eriner, thoughts?

**Comment by protist** [comment]
@ericbn That looks ideal from your example, but it doesn't seem to work for me. It still has the same behaviour as prior to changing that setting. i.e. it still fully completes the word.
```
$ setopt COMPLETE_IN_WORD
$ touch bar<tab>
$ touch foo-bar<tab>
           ^ cursor here
$ touch foo-bar
                ^ cursor here
```

**Comment by protist** [comment]
Thank you @ericbn. That commit works perfectly! Now the behaviour is as follows.
```
$ touch bar<tab>
$ touch foo-bar<tab>
              ^ cursor here
$ touch foo-bar
              ^ cursor here
-- file --
foo-bar      foo-baz-bar
```
And a subsequent tab can select from the menu. This is great, and no longer automatically completes. Just for the record, the cursor position was slightly unintuitive at first. Previously, it was at the `-`, and now it's at the `r`. The latter is the first ambiguous character, so it makes sense. I guess previously zsh considered `-` to act more as a word break, but now I think about it, the current behaviour makes more sense to me. I think it's better like this, agnostic of what a "word" is.

Thanks again!

**Comment by ericbn** [comment]
@protist, I also think it behaves better now. But having just fuzzy matching is giving me too many matches, as I usually want to match from the beginning of the name.

So maybe having zsh first try to match from the beginning, and then (if there's no match) try to do the fuzzy match is a better configuration. All case insensitive, of course.

Agree? @Eriner, thoughts?

**Comment by protist** [comment]
I haven't really noticed anything annoying yet personally, but I think I probably agree with you conceptually. Presumably, if you have `barty` and `foobar` in a directory, then typing `bar` or even `b` should complete to the former. That does make sense to me.

**Comment by ericbn** [comment]
Done. Thanks again for your input, @protist.

**Comment by protist** [comment]
No worries @ericbn. Thank you for the quick fixes and responsiveness!

Unfortunately though, the behaviour of your latest commit is essentially back to where we started!
```
$ touch foo-bar foo-baz-bar
$ touch bar<tab>
$ touch foo-bar<tab>
              ^ cursor here
$ touch foo-bar
                ^ cursor here
```

The second `tab` gives no option to complete, unlike zim's [previous behaviour here](https://github.com/zimfw/zimfw/issues/212#issuecomment-375450712). I can see the internal logic; it's consistent with [your comments](https://github.com/zimfw/zimfw/issues/212#issuecomment-375669803) to "first try to match from the beginning". I'm not really sure what the resolution should be. The behaviour after the first `tab` makes sense. In isolation (with the first `tab`'s behaviour hidden), so does the second. Perhaps one way to differentiate our expected behaviour might be to use the position of the cursor before the second `tab` is pressed. Is that possible/reasonable?

**Comment by AtomicCoding** [comment]
It works for me: [(asciinema recording)](https://asciinema.org/a/kZHLZGKUV3BXiCxpaodnkEeUU). I'm on the latest commit.

**Comment by protist** [comment]
@AtomicCoding I don't think you are testing the same thing, but it's hard to tell without knowing your directory's contents. In your example, create another file called `mafoobarin.py` first, then try the tab completion of `in.py` again, pressing tab twice. It will only auto-complete `main.py`. That's what this issue is referring to.

**Comment by AtomicCoding** [comment]
@protist Sorry! Misunderstood!

**Comment by protist** [comment]
@AtomicCoding No worries. Thanks for the interest though!

**Comment by AtomicCoding** [comment]
@protist It worked [here](https://asciinema.org/a/CWngOwb0sGdPvoV6NCsgo3UM9).

**Comment by ericbn** [comment]
@AtomicCoding, I cannot reproduce your last video recording with `touch mafoobarin.py main.py<enter>touch in.py<tab><tab>`. What you have there is the desired behavior indeed, but that's not what I'm getting yet.

My zim is also updated to master (13eb86d), and I tried it with zsh 5.3 and zsh 5.5.

**Comment by AtomicCoding** [comment]
@ericbn @protist 
These are my zsh options:
```
alwaystoend
autocd
autopushd
autoresume
nobgnice
nocaseglob
nocheckjobs
noclobber
completeinword
correct
extendedglob
extendedhistory
globcomplete
histignorealldups
histignoredups
histignorespace
histsavenodups
histverify
nohup
incappendhistory
interactive
interactivecomments
nolistbeep
login
longlistjobs
menucomplete
monitor
pathdirs
promptsubst
pushdignoredups
pushdsilent
pushdtohome
sharehistory
shinstdin
zle
```
Maybe try running:
```zsh
setopt alwaystoend autocd autopushd autoresume nobgnice nocaseglob nocheckjobs noclobber completeinword correct extendedglob extendedhistory globcomplete histignorealldups histignoredups histignorespace histsavenodups histverify nohup incappendhistory interactive interactivecomments nolistbeep login longlistjobs menucomplete pathdirs promptsubst pushdignoredups pushdsilent pushdtohome sharehistory shinstdin
```

**Comment by Eriner** [comment]
@AtomicCoding What happens if you move `completion` to be the last module loaded?

**Comment by AtomicCoding** [comment]
@Eriner Still works. I did some testing and it seems to be my custom plugin `brobot`. The only thing in the plugin that could affect it is `setopt EXTENDED_GLOB GLOB_COMPLETE MENU_COMPLETE`. So it seems to be that one of those isn't set, but it's likely `MENU_COMPLETE`.

**Comment by AtomicCoding** [comment]
@Eriner @ericbn @protist Can you try setting `setopt GLOB_COMPLETE` and trying again?

**Comment by ericbn** [comment]
Doing `setopt GLOB_COMPLETE` works, but I consider that a side effect, as `GLOB_COMPLETE` is not what we want to do here.

**Comment by ericbn** [comment]
I might have fixed this. @protist, please check if this works for you (replacing the matcher-list in modules/completion/init.zsh):

    zstyle ':completion:*' matcher-list 'm:{a-zA-Z}={A-Za-z} l:|=*' '+r:|?=**'

What this does is that it anchors the match from the left to the beginning of the string in the command line.

**Comment by AtomicCoding** [comment]
@ericbn Works for me.

**Comment by protist** [comment]
I didn't totally understand all @AtomicCoding's comments, so I haven't tested that yet. Let me know if you would like me to go back and test.

@ericbn, testing your fix partially works. For the following, it is fine:
```
$ touch main.py mafoin.py
$ touch in.py<tab>
$ touch in.py<tab>
        ^ cursor moves here
-- file --
mafoin.py  main.py  
```
It mostly works after testing by adding a filename starting with the typed portion.
```
$ touch in.pyfoo
$ ls
in.pyfoo  mafoin.py  main.py
$ touch in.py<tab>
$ touch In.py<tab>  # note the change in case of the initial i???
             ^ cursor remains here
$ touch In.py
             ^ cursor remains here
-- file --
in.pyfoo   mafoin.py  main.py  
```
No idea where the upper-case `I` came from. There are no upper-case characters at all here.

Also, our earlier test fails too.
```
$ rm *; touch foo-bar foo-baz-bar
$ touch bar<tab>
$ touch or-bar<tab>
        ^ cursor moves here
$ touch or-bar
              ^ cursor moves here
-- no matches found --
```

**Comment by ericbn** [comment]
Thanks for testing this, @protist!

I see now that the culprit is the case-insensitive pattern there. If you use

    zstyle ':completion:*' matcher-list 'l:|=*' '+r:|?=**'

all your tests above should pass. But we would miss the case-insensitivity, so I'm still struggling with this `matcher-list` configuration...  Anyone expert on this around?  :- )

**Comment by protist** [comment]
No worries! Thank you for your continued work on this @ericbn!

I can confirm that the three tests work on my system too. And also that the case-insensitivity is lost.

However, aren't we basically back to your comments [here](https://github.com/zimfw/zimfw/issues/212#issuecomment-375669803), where completion doesn't prioritise matching from the beginning of the name? (Given the complications, perhaps this is impossible, so maybe we'd have to concede that this is as good as it gets?)

**Comment by ericbn** [comment]
@protist, yeah. It works when we remove one of these two:
* [match from left first](https://github.com/zimfw/zimfw/commit/98eef4184a7298a6f27388028dbfd4a5ff4b5ff6)
* [case-insensitivity](https://github.com/zimfw/zimfw/issues/212#issuecomment-380817483)

but it gets weird when we have both. And both are important I think...  :- /

**Comment by protist** [comment]
@ericbn Was this meant to be closed? The referenced commit was an old one, right? And has subsequently been modified anyway?

**Comment by ericbn** [comment]
I think this still needs to be fixed, and that the current configuration we have is not good enough.

**Comment by ericbn** [comment]
@Eriner, I'm divided here after the long discussion that happened around this issue.

I'm **not** happy with the current `matcher-list` config we have in our completion module. I've gathered some nice configurations that work. Can you help me choose?

* `zstyle ':completion:*' matcher-list 'm:{a-zA-Z}={A-Za-z} r:|[.,_-]=** r:|=*'`
    Completes f-b to foo-bar (or to foo-something-bar) or anything with the separators above.
* `zstyle ':completion:*' matcher-list 'm:{a-zA-Z}={A-Za-z} l:|=* r:|=*'`
    Completes with any substring match, in the beginning, middle or end of the trial completion strings. Has the weird behavior of completing bar to or-bar, when there are foo-bar and foo-baz-bar files.
* `zstyle ':completion:*' matcher-list 'm:{a-zA-Z}={A-Za-z} r:|?=** r:|=*'`
    Fuzzy match, so even fb can match foobar.

All above are case-insensitive. Only the first one matches from the beginning first. I wish I could make the last one to that, but I could not.

... Or maybe something else that works?

**Comment by edwardsmit** [comment]
Does there need to be one chosen? Can you maybe make this configurable? Along with this excellent explanation the end user can then make his own decision.

**Comment by protist** [comment]
> Can you maybe make this configurable?

@edwardsmit You can just put the relevant `zstyle` line in your `zshrc`, after the `zim` initiation, of course. Only the last definition is used. (You can check this with `zstyle -L | grep matcher`.)

@ericbn I've actually lost track a bit of which configs we've tried before. I tested the suggestions in a new directory after `touch foo-bar foo-something-bar`, and basically tried to break each one. 😬 My bullet points correspond with each bullet point of code from your last comment, in the same order.

* `touch bar<TAB>` does nothing.
* `touch f-b<TAB>` does nothing. (Also, oddly enough `touch bar<TAB>` completed to `-bar` for me with my two files, whereas with the two files you mention, I get the same behaviour as you.)
* This one seemed the most robust, but it's liberalness can obviously be a little off-putting at time. Also, I could finally make it break with `foo-bar` and `foo-baz-bar` as follows:
```
$ touch bar<tab>
$ touch or-bar<tab>
        ^ cursor moves here
$ touch or-bar
              ^ cursor moves here
-- no matches found --
```
(Actually, this fails similarly to some other code previously in this thread.)

**Comment by AtomicCoding** [comment]
@ericbn 
[![asciicast](https://asciinema.org/a/HsBqmUcEz9aOxSQiidLKpknje.png)](https://asciinema.org/a/HsBqmUcEz9aOxSQiidLKpknje)
Works for me!

**Comment by protist** [comment]
> I found an issue in unix.stackexchange exactly like yours

Huh, what a startling coincidence! 😲

However, I tried all three variants, but for me they still fail as per the original post. Firstly, `touch foo-bar foo-something-bar`, then:
```
$ touch bar<tab>
$ touch foo-bar<tab>
           ^ cursor here
$ touch foo-bar
                ^ cursor here
```

**Comment by ericbn** [comment]
@protist, I get mixed results depending on if I try to complete from `touch bar<tab>` or `vi bar<tab>`, and if I'm on Mac or PC:

Given we're on an empty directory and then `touch foo-bar foo-something-bar`.

zstyle | `touch bar<tab>` | `vi bar<tab>`
---|---|---
`'r:\|=*' 'l:\|=* r:\|=*'` | Good on Mac, **Bad** on PC | Good on Mac and PC
`'r:\|=*' '+l:\|=*'` | Good on Mac, **Bad** on PC | Good on Mac and PC
`'m:{a-zA-Z}={A-Za-z} r:\|=*' '+l:\|=*'` | **Bad** on Mac and PC | Good on Mac and PC

By "Good" I mean offering both files in the completion menu. By "Bad" I mean completing to `foo-bar` without showing the completion menu.

Tried with zsh 5.3 and zsh 5.5.1 running on MacOS 10.13.4 (Mac), and zsh 5.4.2 running on Ubuntu 16.04.3 on Linux Subsystem for Windows (PC).

**Comment by protist** [comment]
@ericbn Thanks for that extremely thorough troubleshooting! I'm on Linux with zsh 5.5.1. I have no idea why the same version of zsh would produce different results on Mac and Linux. Seems very odd.

I also tried the first zstyle with `vi bar<tab>`, and it worked too, kinda. I get partial completion to `-bar` (I expected completion to `foo-bar`?), then the completion menu. Do you see completion to `-bar` for both `touch bar` and `vi bar`, or does the former complete to `foo-bar` as per my posts above?

**Comment by ericbn** [comment]
@protist, the partial completion to `-bar` (instead of `foo-bar` with the cursor in the middle) is actually what I think solves the main issue -- that is not having the completion menu (because there's a file that is named `foo-bar`). I also think it's a simpler behavior than having the completion from both sides at the same time, so I like it.  :- )

I get all kinds of different completions depending of the OS or zsh version I'm in. I just classified as "Bad" when I don't get the completion menu, and "Good" when I do get it (no matter how many <kbd>tab</kbd>s I have to press).

**Comment by protist** [comment]
@ericbn Sorry for the delay; I've been quite ill. Your explanation makes sense. I actually don't feel strongly about how much gets completed, as long as the completion menu appears after. The only reason I mentioned it was because I thought it might be diagnostic for why/when "Bad"ness and "Good"ness occurs. I also find it bizarre that this difference exists between Mac and Linux.

**Comment by protist** [comment]
Thanks @ericbn! Tested and works well. I went through the same [testing above](https://github.com/zimfw/zimfw/issues/212#issuecomment-383283006), and it all works as expected. Thank you for persisting and well done on fixing this!

**Comment by Eriner** [comment]
Wow, cheers @ericbn and @protist. What a trip.

**Comment by protist** [comment]
Oh no @ericbn! Sorry, it's still buggy.

[From above](https://github.com/zimfw/zimfw/issues/212#issuecomment-375669803), I think we agreed that zsh should attempt to match from the beginning of the word first, and only if that failed to use the "fuzzier" matching. However, this doesn't seem to be the case.
```
$ rm *
$ touch baritone fboaor
$ touch b<tab>
$ touch bar<tab>
```
The first tab only completes to bar, and the second tab will show the selection menu. I think our previous thoughts were that the first tab should immediately complete to `baritone`?

**Comment by ericbn** [comment]
Just reverted the commit. Back to the drawing board...

**Comment by protist** [comment]
I've been coming across weird ones every now and then. This was a particularly weird one today.

```
$ mkdir test; cd test
$ touch BagFrame.gcode Cats+ScorePad.gcode CatToken_Rumi.gcode
$ ls -1
BagFrame.gcode
Cats+ScorePad.gcode
CatToken_Rumi.gcode
$ ls <tab>
$ ls ca<tab>
$ ls CatT<tab>
$ ls CatToken_Rumi.gcode
```

The word starting with `B` isn't even an option.

**Comment by ericbn** [comment]
> I've been coming across weird ones every now and then. This was a particularly weird one today.

Hi @protist. Looks like this is an issue with Zsh 5.9 when using `m:{a-zA-Z}={A-Za-z}` in the completion matcher-list, which our completion module does use. See https://www.zsh.org/mla/workers/2022/msg01229.html

**Comment by protist** [comment]
Thanks @ericbn. Nice find. I tested the patch, and it does indeed fix that latest issue.

**Comment by ericbn** [comment]
@protist, thanks for the feedback. Good point, I agree the partial completion is a nice to have. Both options I proposed were getting rid of them.

Not having partial completions fixes some of issues here, since the completion system won't have to pick up from those partial completions. If only we could make them cleaner.

**Comment by ericbn** [comment]
I've given up on the "way too fuzzy" matcher-list being used before. There's just one scenario that is off when case is insensitive, but I'm assuming this is the Zsh bug related to https://www.zsh.org/mla/workers/2022/msg01229.html

setup:
```zsh
% touch something_else USER
```
good:
```zsh
% ls se<TAB>
-- inputs --
something_else USER
```

good:
```zsh
% ls sE<TAB>
-- inputs --
something_else USER
```

weird:
```zsh
% ls Se<TAB>
% ls USER
```

weird:
```zsh
% ls SE<TAB>
% ls USER
```

**Comment by protist** [comment]
Thanks so much for persisting on this @ericbn .

I've lost track a bit of what we tried and what failed, so I might be running the wrong tests. Regardless, I tested the version just before and after this commit. My original [issue](https://github.com/zimfw/zimfw/issues/212#issue-2558977800 is unfortunately not fixed. Prior to this commit
```
$ touch foo-bar foo-baz-bar
$ touch bar<tab>
$ touch foo-bar<tab>
              ^ cursor here
$ touch foo-bar
                ^ cursor here
```

and after the commit
```
$ touch foo-bar foo-baz-bar
$ touch bar<tab>
$ touch foo-bar<tab>
           ^ cursor here
$ touch foo-bar
                ^ cursor here
```
Both fail to allow selecting `foo-baz-bar`, although the position of the cursor after the first tab is slightly different.

Also, I can replicate your "weird" situations starting from  `% ls Se<TAB>`  after the newest commit. However, when testing the version before the commit, there is no problem. This behaves as you would expect.

**Comment by protist** [comment]
I'm so sorry @ericbn . I feel terrible.

Also, this is kinda weird, but on my system where I updated to the latest commit 987823e, I can no longer autocomplete the middle of a word. e.g.

```
$ touch foo-abc-bar
$ touch foo-bar<tab>
-- no matches found --
```

On my other, non-updated system running c131777 (3 commits earlier), this completes fine. However, when I revert my broken updated system to this same commit, then open a new terminal, it still fails 🤷 No idea how that could work.

**Comment by ericbn** [comment]
Hi @protist, the extra fuzziness was removed, so you can do `touch foo-a-b<tab>`, but not `touch foo-bar<tab>` anymore.

I feel I need to create a table with all the different test scenarios, all the different matcher-list configurations, and how many pass or fail. Maybe even start by taking out the case insensitive match, which on itself is another layer of complexity. We've even seen different behaviors of the same option but in different platforms (PC vs Mac) in the past, so this mind-blowingly complex overall.

On one end, the most fuzzy option is just `'r:|?=**'` (or `'m:{[:lower:][:upper:]}={[:upper:][:lower:]} r:|?=**'`).

The option used by prezto is `'r:|[._-]=* r:|=*' 'l:|=* r:|=*'` ([here](https://github.com/sorin-ionescu/prezto/blob/6e564503f1c5e6ddba2bcf5d9065e5872ca207d2/modules/completion/init.zsh#L84)) and by ohmyzsh is `'r:|=*' 'l:|=* r:|=*'` ([here](https://github.com/ohmyzsh/ohmyzsh/blob/fa396ad7718af77e896472772bffd4a7e9110193/lib/completion.zsh#L18)).

I went with almost the same as prezto for case insensitive: `'' 'r:|[._-]=* r:|=*' 'l:|=* r:|=*'` ([here](https://github.com/zimfw/completion/blob/879dddb33c3434429277134dc546a891873a5dc4/init.zsh#L88), and for case insensitive a variation of that which is more case insensitive than prezto's variation.

I see that having more than one item in the matcher-list does create a mess when typing <tab> more than once.

On the other hand, being too fuzzy, like the option I've mentioned above, does create a lot of options. A test case by @cattyhouse ([here](https://github.com/zimfw/completion/commit/e0306adc653d682d6fcb32049418cbb9fa8a0d43#commitcomment-69250971)) is just typing `mv<tab>`.

A simple option is just `'l:|=* r:|[._-]=*'`. It will complete from beginning, middle or end, and will also expand terms separated by `[._-]`, so `touch f-a-b<tab>` will expand to `touch foo-abc-bar`. We don't have any test cases with this type of expansion as far as I've checked. An even simpler options is just `'l:|=* r:|=*'`.

Taking the other frameworks as comparison, looks like ohmyzsh users do not care about any fuzziness at all. Prezto does add the `[._-]` separators completion and uses `_approximate` to add some "fuzziness" in terms of allowing some typing errors. Maybe I went to far with the fuzziness with the options I've used, which have been intentionally deviating from prezto or ohmyzsh.

**Comment by protist** [comment]
> the extra fuzziness was removed, so you can do `touch foo-a-b<tab>`, but not `touch foo-bar<tab>` anymore.

Ah right. I actually came across the new behaviour in a common scenario that I use.

1. Have one original file that want to rename, say `form.docx`
2. Rename it with [renameutils](https://www.nongnu.org/renameutils/), i.e. `imv form.docx`. Let's say the new file is now named `form.signed.docx`.
3. Attempt to access the file on the CLI. e.g.
    * `touch <ALT+.>`, which gives
    * `touch form.docx`
    * Then try to tab complete.

In the past this would complete to `form.signed.docx`, but now doesn't. So for me at least, this level of fuzziness was helpful!

> I feel I need to create a table

That sounds like a great idea. I've lost track of all the permutations myself! Thanks for the details. There are so many, and it's definitely tricky to find the ideal level of fuzziness.

**Comment by ericbn** [comment]
@protist, regarding your `touch form.docx<tab> -> touch form.signed.docx` example:

1. It didn't work like that before with our `zstyle ':completion:*' matcher-list '' 'r:|?=**'` configuration.
2. Found a counter example below that I find annoying (`|` is the cursor position):

```zsh
% git rebase --interactive --autosq|<tab>
% git rebase --interactive --|autosquash<tab>
-- option --
--autosquash     -- check for auto-squash boundaries
--no-autosquash  -- do not check for auto-squash boundaries
# Subsequent tabs will toggle between --autosquash and --no-autosquash infinitely.
```

This is also a counter example for what I've said before which I take back: "I see that having more than one item in the matcher-list does create a mess when typing `<tab>` more than once." Again, it's not that simple.

**Comment by protist** [comment]
> 1. It didn't work like that before

Oh weird. I still have an old system that is doing it they way I expected. And I'm *fairly* sure that my updated system changed after the update. IDK what's going on; maybe completion is defined elsewhere for me.

> 2. Found a counter example

Ouch, it really is confusing.

**Comment by ericbn** [comment]
@protist I believe I've narrowed down what the issue is. I've reported it in https://www.zsh.org/mla/workers/2025/msg00125.html

**Comment by ericbn** [comment]
I'm happy with the solution we have in place, that has a good balance between being fuzzy (not as fuzzy as before and as discussed here) and being consistent (it was buggy before). This was done [here](https://github.com/zimfw/completion/compare/8a73f55e7ab0077ed34bfc3cb4db4e71afb7d952...8321237c096e67822fcb666fa60fa01ee60ce1d7).

Closing this issue.
